modules/ndu/src/backend/ndu-engine.ts

```diff
@@ -26,49 +68,70 @@ export class UnderstandingEngine {
-    //       suggestion.decision.status = 'dropped'
-    //       suggestion.decision.reason = 'would have been elected, but already in the middle of a flow'
-    //     }
-    //   })
-    // }
+    debug('Processing %o', { currentFlow, activeContexts, isInMiddleOfFlow })
 
-    await this._processTriggers(event)
-
-    const elected = event.suggestions!.find(x => x.decision.status === 'elected')
     const actions = []
 
-    if (elected) {
-      // QNA are always active
-      const payloads = _.filter(elected.payloads, p => p.type !== 'redirect')
-      if (payloads) {
-        actions.push({ action: 'send', data: elected })
-      }
+    // Need some magic to get the best topic (maybe it's the second one with better scoring intents? )
+    const bestTopic = _.findKey(event.nlu.predictions, x => x.confidence > this.MIN_CONFIDENCE)
 
-      // Ignore redirections when in middle of flow
-      if (!isInMiddleOfFlow) {
-        const redirect = _.find(elected.payloads, p => p.type === 'redirect')
-        if (redirect && redirect.flow && redirect.node) {
-          actions.push({ action: 'redirect', data: redirect })
-          actions.push({ action: 'continue' })
-        }
+    // No active goal or contexts, so we'll add the top as context
+    if (!(currentFlow && isInMiddleOfFlow) && !activeContexts.length && event.nlu.predictions) {
+      if (bestTopic) {
+        event.state.session.nluContexts = [...(event.state.session.nluContexts || []), { context: bestTopic, ttl: 3 }]
+        debug(`No active goal or context. Activate topic with highest confidence: ${bestTopic} `)
+      } else {
+        debug(`No active goal or context. Top context prediction too low `)
       }
     }
 
-    // When not actively in a flow and a trigger is active, redirect the user
-    const completedTriggers = this._getGoalsWithCompletedTriggers(event)
-    if (completedTriggers.length && !isInMiddleOfFlow) {
-      const [topic] = completedTriggers[0].split('/')
+    const predictionScores = this.getPredictionScores(event.nlu.predictions)
+
+    const bestIntents = event.nlu.predictions?.[bestTopic]?.intents?.map(x => ({
+      name: x.label,
+      confidence: x.confidence,
+      context: bestTopic
+    }))
+
+    // Overwrite the NLU detected intents
+    event.nlu.intent = bestIntents?.[0]
+    event.nlu.intents = bestIntents
+
+    // Then process triggers on what the NDU decided
+    await this._processTriggers(event)
+
+    const possibleGoals = this._getGoalsWithCompletedTriggers(event)
+    possibleGoals.length && debug(`Possible goals: %o`, possibleGoals)
+
+    // If it's a QNA, we query the module to get the actions to execute
+    if (event.nlu.intent?.name?.startsWith('__qna__')) {
+      debug(`Sending knowledge for %o`, event.nlu.intent)
+      const qnaActions = await this.queryQna(event.nlu.intent.name, event)
+
+      // if (isInMiddleOfFlow && qnaActions.find(x => x.action === 'redirect')) {
```

Comment:
Dead code?

Reply:
Unsure if we keep it or not, so leaving it for now until we work out the final logic of the NDU